src/server/pkg/obj/obj.go

```diff
@@ -222,8 +222,57 @@ type Client interface {
 	IsIgnorable(err error) bool
 }
 
+type checkedReadCloser struct {
+	io.ReadCloser
+	size  uint64
+	count uint64
+}
+
+func newCheckedReadCloser(size uint64, rc io.ReadCloser) io.ReadCloser {
+	return &checkedReadCloser{
+		ReadCloser: rc,
+		size:       size,
+	}
+}
+
+func (crc *checkedReadCloser) Read(p []byte) (int, error) {
+	count, err := crc.ReadCloser.Read(p)
+	crc.count += uint64(count)
+	if err != nil {
+		if errors.Is(err, io.EOF) && crc.count != crc.size {
+			return count, errors.Errorf("read stream ended after the wrong length, expected: %d, actual: %d", crc.size, crc.count)
+		}
+		return count, err
+	} else if crc.count > crc.size {
```

Comment:
I think we might want to restructure this so that this check happens before the `err != nil`. We could hit a situation where it is an error to read too many bytes, and whatever error message is generated by the object storage client will be returned rather than the error message in the checked reader which would tell us we read too many bytes. It could be even better to actually wrap the object storage error in the checked reader error message.

Reply:
I'm not very concerned about an object storage client erroring if we read too many bytes, as we would error even without that, and that implies a pretty flagrant bug on the underlying object client, so I'd rather have the underlying error.  I think wrapping the error is a good thing to do in the first `return`, though, so I'll add that.

Reply:
I left this check here but added the same check in the case of an error because we want to do different things in each branch - wrap an error or return a new one.